### DISCUSSION TITLE
How to exclude null value from pydantic model used by FAST API

### DISCUSSION THREAD

**Discussion by karan-kumar-by** [fastapi/fastapi#8882] [Questions] [answered]
I am currently using pydantic model as below. It has 2 optional fields description and tax. 

from typing import Optional
from fastapi import Body, FastAPI
from pydantic import BaseModel, Field

app = FastAPI()

class Item(BaseModel):
    name: str
    description: Optional[str] = Field(None, title="The description of the item", max_length=300)
    price: float = Field(..., gt=0, description="The price must be greater than zero")
    tax: Optional[float] = None

@app.post("/items/")
async def create_item(item: Item):
    return item

**Problem -** How to exclude optional fields from JSON when not supplied as part of the request.

Request -:
 {
 	"name": "A",
 	"price": 10.21
 }

Fast API produces -:
{
  "name": "A",
  "description": null,
  "price": 10.21,
  "tax": null
}

Expected -:
 {
 	"name": "A",
 	"price": 10.21
 }

PYDANTIC provides a way (exclude_none flag) to do that but how to make use of that flag when it is being used in fast API?

**Comment by karan-kumar-by** [comment]
This what pydantic provides - https://pydantic-docs.helpmanual.io/usage/exporting_models/#modeldict

**Comment by gyKa** [comment]
Try this code:

```python
@app.post("/items/")
async def create_item(item: Item):
    return item.dict(exclude_none=True)
```

**Comment by plankthom** [comment]
Using `exclude_none` doesn't seem to work when a `response_model` is in mentioned in the route decorator.

```python
@app.post("/items/", response_model=Item)
async def create_item(item: Item):
    return item.dict(exclude_none=True)
```
Fast api seems to reprocess the dict with the pydantic model

So i settled for overriding the `dict` method in the model itself
```python
def Item(BaseModel):
   name: str
   description: Optional[str]
   ...
   def dict(self, *args, **kwargs) -> Dict[str, Any]:
        _ignored = kwargs.pop('exclude_none')
        return super().dict(*args, exclude_none=True, **kwargs)
```
(an actual solution would put this definition in separate subclass of `BaseModel` for reuse)

Note: just changing the default value of the. `exclude_none` keyword argument is not enough: it seems FastAPI always sends `exclude_none=False` as an argument ...

**Comment by squat** [comment, marked as answer]
You can accomplish this directly with the decorator by setting the `response_model_exclude_none` parameter, e.g.:

```python
@app.post("/items/", response_model=Item, response_model_exclude_none=True)
async def create_item(item: Item):
    # do something with item
    return item
```
This was added last year in https://github.com/tiangolo/fastapi/pull/1166 and is available since version 0.54.0

**Reply by domisum** [reply]
Make sure to include the `response_model=` as well, otherwise the exclude flag won't take effect

**Comment by karan-kumar-by** [comment]
@squat I need to exclude None from the request body, not from the response model.

`from typing import Optional
from fastapi import Body, FastAPI
from pydantic import BaseModel, Field
from typing import List

app = FastAPI()


class Item(BaseModel):
    name: str
    description: Optional[str] = Field(None, title="The description of the item", max_length=300)
    price: float = Field(..., gt=0, description="The price must be greater than zero")
    tax: Optional[float] = None


@app.post("/items/")
async def create_item(item: List[Item] = Body(...)):
    return item

`
Request -:
{
"name": "A",
"price": 10.21
}

Fast API produces -:
{
"name": "A",
"description": null,
"price": 10.21,
"tax": null
}

Expected -:
{
"name": "A",
"price": 10.21
}

Question is - How to send the exclude_none or exclude_default flag to the PYDANTIC model?

**Reply by rafrafek** [reply]
Did you find any answer to this?

**Reply by YuriiMotov** [reply]
Seems that the author needs `json_data = item.model_dump(exclude_unset=True)`

**Reply by rafrafek** [reply]
Thanks, but where to put the `json_data`?

My code:

```python
from fastapi import FastAPI, Response
from pydantic import BaseModel

class Schema(BaseModel):
    foo: str
    bar: str | None = None

app = FastAPI()

@app.post("/process")
async def process(item: Schema) -> Response:
    print(item)
    return Response()

```

It produces a schema that says `bar` can be `null`, but I don't want it to be `null`. It has to be either `string` or not provided in the request at all.

<img width="360" alt="image" src="https://github.com/user-attachments/assets/f748e47a-245d-4bc1-becf-cc3568fe8085">

I just want it to be a non-required string. So `string` but without the star `*`.

I think the issue is that Pydantic does not support per-field `omit_if_none`/`omit_if_default` settings https://github.com/pydantic/pydantic/discussions/5461.
And it's easier in JavaScript because Python doesn't support `undefined` value, only `null` (`None`).

**Reply by YuriiMotov** [reply]
You can use `str | SkipJsonSchema[None] = None` annotation. This allows the parameter  to be None, but excludes null from schema:

```
from pydantic.json_schema import SkipJsonSchema

class Schema(BaseModel):
    foo: str
    bar: str | SkipJsonSchema[None] = None
```

But this way, the request with `bar=null` is valid.

You can also use this approach:
```
class Schema(BaseModel):
    foo: str
    bar: str = "not specified"
```

**Reply by rafrafek** [reply]
Thank you! With `str | SkipJsonSchema[None] = None` it works:

<img width="254" alt="image" src="https://github.com/user-attachments/assets/b164e2ff-7724-447e-82b0-81569ea90ec5">

The field looks perfect now!

Is there a way to apply `SkipJsonSchema` to all optional fields, or does it have to be done manually for each field in every model?

**Reply by michaeloliverx** [reply]
An option to apply this to all optional fields would be useful, I am using an autogenerated client in my frontend and all optional fields are typed as `type | undefined | null` where I'd like it to be only typed as `type | undefined`.

**Comment by squat** [comment]
Your initial question was
>How to exclude optional fields from JSON

However, you're saying that you're _not_ concerned with the JSON response from the API handler, correct?

 I don't understand what you're trying to accomplish. JSON is only relevant in two places here: the serialized request and the serialized response. You can't control the whether or not a field is present in a request's JSON because that comes from the client. Are you worried about JSON in requests coming from clients generated by your OpenAPI spec? Again, this is something out of FastAPI's control.

Maybe it would be helpful if you share _when_/_where_ you expect this? e.g.:
* do you want JSON to exclude these fields? As mentioned above you can only enforce the response;
* if it's not about JSON, do you want the Python `item` object _inside_ the handler to exclude these fields? `exclude_none` is used when turning a Pydantic object into a dict, e.g. right before the handler returns JSON. `item` is not a dict; it is a Python object of a type that you defined as having a `description` and `tax` field. These fields will _always_ be present on the `item` object, regardless of whether the request JSON had them. Remember, `Optional` is a bit of a misnomer; it does not mean that having the field is optional but rather that the _value_ can be `None`. It is defined as shorthand for a union between `None` and another type.

**Comment by karan-kumar-by** [comment]
@squat Thanks for the detailed explanation. 

Let me ask this way - How to exclude none before the handler returns JSON?

**Comment by AthreyVinay** [comment]
@plankthom Your solution works for a get operation. On a put, I get the following:
```python
KeyError: 'exclude_none'
```
Could you kindly confirm? Thanks

**Comment by guybartal** [comment]
+1

**Comment by daipham3213** [comment]
> @plankthom Your solution works for a get operation. On a put, I get the following:
> 
> ```python
> KeyError: 'exclude_none'
> ```
> 
> Could you kindly confirm? Thanks

You could check if the attr exist before remove it

```python
def Item(BaseModel):
   name: str
   description: Optional[str]
   ...
   def dict(self, *args, **kwargs) -> Dict[str, Any]:
         if hasattr(kwargs, "exclude_none"):
            _ignored = kwargs.pop("exclude_none")
        return super().dict(*args, exclude_none=True, **kwargs)
```

**Comment by AthreyVinay** [comment]
> > @plankthom Your solution works for a get operation. On a put, I get the following:
> > ```python
> > KeyError: 'exclude_none'
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Could you kindly confirm? Thanks
> 
> You could check if the attr exist before remove it
> 
> ```python
> def Item(BaseModel):
>    name: str
>    description: Optional[str]
>    ...
>    def dict(self, *args, **kwargs) -> Dict[str, Any]:
>          if hasattr(kwargs, "exclude_none"):
>             _ignored = kwargs.pop("exclude_none")
>         return super().dict(*args, exclude_none=True, **kwargs)
> ```

hi @daipham3213 

I tried your solution with the following error:
```python
TypeError: pydantic.main.BaseModel.dict() got multiple values for keyword argument 'exclude_none'
```

**Comment by arrunpras** [comment]
if someone still looking for this answer, this worked for me

```
@app.post("/items/") 
async def create_item(item: Item): 
    return item.dict(exclude_unset=True)
```

**Reply by greatvovan** [reply]
@squat's suggestion looks great, but I found this solution by @arrunpras (and derived from that `response_model_exclude_unset` router parameter) more useful, because you often want to have runtime control for which properties to include, and by leaving some properties unset in the model initializer you can achieve exactly that

**Comment by FastestMolasses** [comment]
> > @plankthom Your solution works for a get operation. On a put, I get the following:
> > ```python
> > KeyError: 'exclude_none'
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Could you kindly confirm? Thanks
> 
> You could check if the attr exist before remove it
> 
> ```python
> def Item(BaseModel):
>    name: str
>    description: Optional[str]
>    ...
>    def dict(self, *args, **kwargs) -> Dict[str, Any]:
>          if hasattr(kwargs, "exclude_none"):
>             _ignored = kwargs.pop("exclude_none")
>         return super().dict(*args, exclude_none=True, **kwargs)
> ```

This is a better solution

```python
class Item(BaseModel):
    name: str
    description: Optional[str]
    ...
    def dict(self, *args, **kwargs) -> dict[str, Any]:
        """
            Override the default dict method to exclude None values in the response
        """
        kwargs.pop('exclude_none', None)
        return super().dict(*args, exclude_none=True, **kwargs)
```

**Comment by lucasgadams** [comment]
We really need this as a built in feature. I find myself writing this `dict` override all the time. It should be something we can set either on `Field` or in the `Config`

**Comment by Ceph97** [comment]
For newer Pydantic version this works best:

```
    def model_dump(self):
        """
        This Overrides the default model dump method to exclude None values
        """
        return super().model_dump(exclude_none=True)
```

**Reply by dh-mshedden** [reply]
I can't get this to work the way I expect. Is it supposed to be called by FastAPI when the response is being serialized? I have to call it explicitly, which isn't what I need.

**Comment by ivanvmoreno** [comment]
Agree with the rest of the commenters – I think this use case would be a perfect fit as a `Config` member:

```python
class Profile(BaseModel):
    first_name: str
    last_name: str
    email: Optional[str] = None
    
    class Config:
        none = "ignore"
```

**Comment by Ted-Barrett** [comment]
Here's my brute force **hack** so I don't have to add this to all the decorators and inevitably forget it somewhere:

```python3
def remove_none(d: Dict[Any, Any]) -> Dict[Any, Any]:
    d_without_none: Dict[Any, Any] = {}
    for key, value in d.items():
        if isinstance(value, dict):
            d_without_none[key] = remove_none(value)
        elif isinstance(value, list):
            d_without_none[key] = [
                remove_none(i) if isinstance(i, dict) else i for i in value
            ]
        elif value is not None:
            d_without_none[key] = value
    return d_without_none


# https://dev.to/avirgvd/python-fastapi-middleware-to-modify-request-and-response-body-3f7f
@app.middleware("http")
async def remove_none_middleware(
    request: Request, call_next: Callable[[Request], Awaitable[Response]]
) -> Response:
    response = await call_next(request)

    response_body = b""

    async for chunk in response.body_iterator:  # type: ignore[attr-defined]
        response_body += chunk
    
    non_dict_reponse = Response(
            content=response_body,
            status_code=response.status_code,
            headers=dict(response.headers),
            media_type=response.media_type,
        )

    try:
        response_json = json.loads(response_body.decode("utf-8"))
    except json.JSONDecodeError:
        return non_dict_reponse
    if not isinstance(response_json, dict):
        return non_dict_reponse

    modified_response = json.dumps(remove_none(response_json)).encode("utf-8")
    response.headers["Content-Length"] = str(len(modified_response))
    return Response(
        content=modified_response,
        status_code=response.status_code,
        headers=dict(response.headers),
        media_type=response.media_type,
    )
```

I also modified my `openapi.json` generator to remove the null fields:

```python3
def remove_null(d: Dict[Any, Any]) -> Dict[Any, Any]:
    d_without_null: Dict[Any, Any] = {}
    for key, value in d.items():
        if isinstance(value, dict):
            if value.get("type") != "null":
                d_without_null[key] = remove_null(value)
        elif isinstance(value, list):
            l = []
            for i in value:
                if isinstance(i, dict) and i.get("type") != "null":
                    l.append(i if not isinstance(i, dict) else remove_null(i))
            d_without_null[key] = l
        elif value is not None:
            d_without_null[key] = value
    return d_without_null


if __name__ == "__main__":
    openapi = remove_null(app.openapi())
    version = openapi.get("openapi", "unknown version")

    print(f"writing openapi spec v{version}")
    with open("openapi.json", "w") as f:
        json.dump(openapi, f, indent=2)

```

Note that `remove_none` will **not** remove a `None` which is an element of a list, whereas `remove_null` **will** remove `{"type": "null"}` which is an element of a list.
